### DISCUSSION TITLE
Keeps the time stable for weeks/month?

### DISCUSSION THREAD

**Discussion by basementmedia2** [bxparks/AceTimeClock#9] [General]
Hi Brian, I have again a question about your library (sorry for asking so much;-)
Can it be that the time - once it is set and "kept alive" in the loop with systemClock.loop(); - is gradually adjusted (i.e. after a few days the clock is suddenly e.g. 20 minutes slow?).

**Comment by bxparks** [comment]
(Moving discussion to AceTimeClock project, where the SystemClock classes live since v1.8)

**Comment by bxparks** [comment]
How is the `SystemClockLoop` class configured? More specifically, what is the time source: None (i.e. the internal `millis()`), the DS3231Clock, or the NtpClock?

The only time I've had problems with clock accuracy is due to defective DS3231 chips. I get the impression that the quality of the parts coming out of China through sellers on eBay is not very good. The spec for the DS3231 says that it's supposed to be accurate to 2ppm, so 0.2 seconds per day. But I rarely see a DS3231 chip that accurate. Most of them are off by several seconds per day. The worst one would randomly lose 30-90 minutes after a few days.

If you using the NtpClock, the clock should be accurate to within seconds.

If you are using the internal `millis()`, the internal clock of the microcontroller can be very inaccurate, something like tens of seconds or several minutes per day.

**Comment by basementmedia2** [comment]
Hi Brian,

my current sketch goes like this:

In the setup the ESP32 tries to connect to the WLAN (router). 
If the connection works, the ESP gets the current time via an NTP server and passes it to AceTime.

If the connection does not work, the ESP switches to AP mode and waits for the user to connect to the AP. 
If the user calls the index page of the AP server, the user's device (usually a cell phone) transfers the current time and date to the ESP (via ajax request). These values are then passed to AceTime.

In any case, from then on AceTime should be responsible for obtaining the time, because it may happen that the connection to the router is lost at night (because the router is switched off, for example).

Here is a shortened version of my sketch.
It would be great if you could have a look at it and see if the time stays correct over weeks and months.

If I am wrong here: 
How can I maintain the time with AceTime without internet after the initial setting?

```

#include <WiFi.h>
#include <WiFiClient.h>
#include <WiFiAP.h>
#include <WebServer.h>
WebServer server(80);

#include <AceTime.h>
using namespace ace_time;
using namespace ace_time::clock;

static BasicZoneProcessor pacificProcessor;
static SystemClockLoop systemClock(nullptr /*reference*/, nullptr /*backup*/);
auto pacificTz = TimeZone::forZoneInfo(&zonedb::kZoneEurope_Berlin,&pacificProcessor);

// ++++++++++++++++++++++++++++++++++++++++++ Library for optional initial set of AceTime via internet
#include <time.h> 
const char* ntpServer = "at.pool.ntp.org";
const long  gmtOffset_sec = 3600;
const int   daylightOffset_sec = 3600;
time_t now;
tm tm;

char APssid[25] = "MyAccesspoint";
char APpassword[20] = "123456789";
char ssid[30] = "MySSID";
char password[40] = "MyPassword";
bool timeSet=false;

int currentHour = 0;
int currentMinute = 0;
int currentSecond = 0;
int monthDay = 0;
int currentMonth = 0;
int currentYear = 0;

char formattedTime[10];
char currentDate[20];

const char indexPage[] PROGMEM = R"=====(
  <!DOCTYPE html>
  <html lang="en">
    <script>
    function send_time() {
        var currentdate = new Date(); 
        var currentHour = currentdate.getHours();
        var currentMinute = currentdate.getMinutes();
        var currentSecond = currentdate.getSeconds();
    
        if (currentHour < 10) { currentHour_tmp="0"+currentHour } else { currentHour_tmp=currentHour };
        if (currentMinute < 10) { currentMinute_tmp="0"+currentMinute } else { currentMinute_tmp=currentMinute };
        if (currentSecond < 10) { currentSecond_tmp="0"+currentSecond } else { currentSecond_tmp=currentSecond };
    
    
        var monthDay = currentdate.getDate();
        var currentMonth = (currentdate.getMonth()+1);
        var currentYear = currentdate.getFullYear();
    
        data="currentHour="+currentHour+"&currentMinute="+currentMinute+"&currentSecond="+currentSecond+"&monthDay="+monthDay+"&currentMonth="+currentMonth+"&currentYear="+currentYear;
        console.log(data);
        var url="/update_root?"+data;
        var xmlhttp = new XMLHttpRequest();
        xmlhttp.open('GET', url, true);
        xmlhttp.onreadystatechange = function() {
            if (xmlhttp.readyState == 4) {
                if(xmlhttp.status == 200) {
                    console.log("Date sent");
                }
            }
        };
        xmlhttp.send(null);
    }
    
    send_time();
    </script>
  </html>
)=====";

void setup() {
  Serial.begin(115200);
  
  systemClock.setup();
  
  server.on("/", display_root);
  server.on("/update_root", update_root);

  WiFi.mode(WIFI_AP_STA);

  Serial.print("We try to connect to the Router!");
  WiFi.begin(ssid, password);
  unsigned long pause=10000;
  unsigned long connTimer = millis();
  while (WiFi.status() != WL_CONNECTED && (millis()-connTimer < pause)) {
    Serial.print(".");
    delay(500);
  }
  if (WiFi.status() == WL_CONNECTED) {
    Serial.print("We are connected!");
  } else {
    Serial.print("We are connected so we get back to AP mode.");
    WiFi.mode(WIFI_AP);
  }
  
  WiFi.softAP(APssid, APpassword);

  server.begin(); // Server starten
}


void loop() {
  server.handleClient();
  
  if (timeSet==false) {
    if (WiFi.status() == WL_CONNECTED ) {
      
      // If we are connected to the internet and time is not yet set --> get initial time from internet
      
      time(&now);                       
      localtime_r(&now, &tm);
    
      currentHour = tm.tm_hour;
      currentMinute = tm.tm_min;
      currentSecond = tm.tm_sec;
      monthDay = tm.tm_mday;
      currentMonth = tm.tm_mon + 1;
      currentYear = tm.tm_year + 1900;
           
      auto pacificTime = ZonedDateTime::forComponents(currentYear, currentMonth, monthDay, currentHour, currentMinute, currentSecond, pacificTz);
      systemClock.setNow(pacificTime.toEpochSeconds());
      timeSet=true;
    } 
  } else {
    systemClock.loop();
    acetime_t now = systemClock.getNow();
    
    auto pacificTime = ZonedDateTime::forEpochSeconds(now, pacificTz);
    currentHour = pacificTime.hour();
    currentMinute = pacificTime.minute();
    currentSecond = pacificTime.second();
    monthDay = pacificTime.day();
    currentMonth = pacificTime.month();
    currentYear = pacificTime.year();

    sprintf(currentDate, "%d-%02d-%02d %02d:%02d:%02d", currentYear, currentMonth, monthDay, currentHour, currentMinute, currentSecond);
    Serial.println(currentDate);
  }
  
  delay(1000);
}

void display_root() {
  const char * httpType PROGMEM = "text/html";
  server.send_P(200,  httpType, indexPage);
}

void update_root() {
    while ((currentHour == 0 && currentSecond==0 && currentMinute==0)) {
      for (uint8_t i = 0; i < server.args(); i++) {
        // Wenn die Uhrzeit jetzt noch nicht gesetzt ist, holen wir uns diese vom Handy oder PC
        if ((String)server.argName(i) == "currentHour") {
          currentHour = server.arg(i).toInt();
          Serial.print("Hour set to:");
          Serial.println(currentHour);
        }
        if ((String)server.argName(i) == "currentMinute") {
          currentMinute = server.arg(i).toInt();
          Serial.print("Minute set to:");
          Serial.println(currentMinute);
        }
        if ((String)server.argName(i) == "currentSecond") {
          currentSecond = server.arg(i).toInt();
          Serial.print("Second set to:");
          Serial.println(currentSecond);
        }
        if ((String)server.argName(i) == "monthDay") {
          monthDay = server.arg(i).toInt();
          Serial.print("monthDay set to:");
          Serial.println(monthDay);
        }
        if ((String)server.argName(i) == "currentMonth") {
          currentMonth = server.arg(i).toInt();
          Serial.print("currentMonth set to:");
          Serial.println(currentMonth);
        }
        if ((String)server.argName(i) == "currentYear") {
          currentYear = server.arg(i).toInt();
          Serial.print("currentYear set to:");
          Serial.println(currentYear);
        }
      }
    }

    auto pacificTime = ZonedDateTime::forComponents(currentYear, currentMonth, monthDay, currentHour, currentMinute, currentSecond, pacificTz);
    systemClock.setNow(pacificTime.toEpochSeconds());
    timeSet = true;
}
```

**Reply by bxparks** [reply]
The core of your problem is this line:
```C++
static SystemClockLoop systemClock(nullptr /*reference*/, nullptr /*backup*/);
```

This configures SystemClockLoop to use the internal `millis()` function of the... ESP8266 (?). But the accuracy of the `millis()` function is directly tied to the accuracy of the internal clock of the microcontroller, which may be significantly different from the 80MHz or 160MHz that it is assumed to be.

The hardware solution is to use an external RTC chip, like the DS3231. That's the purpose of the `DS3231Clock`.

The software solution is to synchronize the SystemClockLoop with the NTP client periodically, e.g. every hour. That's the purpose of the `NtpClock`, to be used as the reference clock. I should think about providing a `Clock` implementation that uses the ESP8266 sntp client, but it never got high enough in my priority because my NtpClock class works just fine for me.

A random comment: You might considering renaming `pacificTz` to something like `berlinTz`. I live in America/Los_Angeles time zone, so US Pacific time, and named my variable `pacificTz`.

Second random comment: After you run `time(&now)`, the `now` variable contains the number of seconds from Unix Epoch. To convert this into AceTime Epoch, you simply need to subtract the number of seconds between 1970-01-01T00:00:00 and 2000-01-01-T00:00:00, which happens to be `SECONDS_SINCE_UNIX_EPOCH = 946684800L`. In other words, `acetime_t aceTimeNow = now - SECONDS_SINCE_UNIX_EPOCH`. You don't need to convert `now` into date-time components, then back to `acetime_t`.

**Reply by basementmedia2** [reply]
Thanx for your fast and very helpful answer. 
You Wrote 
"The core of your problem is this line:

static SystemClockLoop systemClock(nullptr /*reference*/, nullptr /*backup*/);
This configures SystemClockLoop to use the internal millis() function of the... ESP8266 (?)"

I use an ESP32.
Can I setup system clock to use the onboard RTC clock?

**Reply by bxparks** [reply]
Maybe, I don't know. The ESP32 apparently has 4 sources of clocks, but I have no idea which one is exposed through its Arduino API. The documentation is scattered, confusing, or nonexistent. It wasn't worth my time to try to figure out the mess. I simply use an external DS3231 chip which is guaranteed to work the same across all processors, and supports backup battery or capacitor. Here are some links that I found in the last 10 minutes that might be helpful for you.

https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/system_time.html
https://github.com/espressif/arduino-esp32/issues/3641
https://www.esp32.com/viewtopic.php?t=3715#p17038
